File: Server/src/models/PokemonInstance.ts
import mongoose, { Document, Schema, Types } from 'mongoose';

export interface IPokemonInstance extends Document {
  owner_id: Types.ObjectId;
  species_id: number;
  nickname: string;
  level: number;
  exp: number;
  nature: string;
  current_hp: number;
  max_hp: number;
  status_condition: string;
  is_in_party: boolean;
  party_slot?: number;
}

const PokemonInstanceSchema: Schema = new Schema({
  owner_id: { type: Schema.Types.ObjectId, ref: 'Player', required: true, index: true },
  species_id: { type: Number, required: true },
  nickname: { type: String, default: '' },
  level: { type: Number, default: 1 },
  exp: { type: Number, default: 0 },
  nature: { type: String, required: true },
  current_hp: { type: Number, required: true },
  max_hp: { type: Number, required: true },
  status_condition: { type: String, default: 'NONE' },
  is_in_party: { type: Boolean, default: false },
  party_slot: { type: Number, min: 1, max: 6 }  // Only required/validated if is_in_party is true
});

// Adding a compound index to quickly find a player's party
PokemonInstanceSchema.index({ owner_id: 1, is_in_party: 1 });

export const PokemonInstanceModel = mongoose.model<IPokemonInstance>('PokemonInstance', PokemonInstanceSchema);

File: Server/Models/PokemonInstance.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace PokemonMMO.Models;

/// <summary>
/// A specific Pokemon owned by a player — optimized with Embedded Documents for Matchmaking.
/// </summary>
public class PokemonInstance
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = null!;

    [BsonElement("owner_id")]
    [BsonRepresentation(BsonType.ObjectId)]
    public string OwnerId { get; set; } = null!;

    [BsonElement("species_id")]
    public int SpeciesId { get; set; }

    [BsonElement("nickname")]
    public string Nickname { get; set; } = "";

    [BsonElement("level")]
    public int Level { get; set; } = 50; // Usually fixed at 50 for Esports

    [BsonElement("exp")]
    public int Exp { get; set; } = 0;

    [BsonElement("nature")]
    public string Nature { get; set; } = null!;

    [BsonElement("current_hp")]
    public int CurrentHp { get; set; }

    [BsonElement("max_hp")]
    public int MaxHp { get; set; }

    [BsonElement("status_condition")]
    public string StatusCondition { get; set; } = "NONE";

    [BsonElement("is_in_party")]
    public bool IsInParty { get; set; } = false;

    [BsonElement("party_slot")]
    [BsonIgnoreIfNull]
    public int? PartySlot { get; set; }

    [BsonElement("is_trial")]
    public bool IsTrial { get; set; } = false;

    [BsonElement("trial_expiry")]
    [BsonIgnoreIfNull]
    public DateTime? TrialExpiry { get; set; }

    [BsonElement("held_item")]
    [BsonIgnoreIfNull]
    public string? HeldItem { get; set; }

    [BsonElement("ivs")]
    public StatBlock Ivs { get; set; } = new();

    [BsonElement("evs")]
    public StatBlock Evs { get; set; } = new();

    [BsonElement("moves")]
    public List<PokemonMove> Moves { get; set; } = new();
}

/// <summary>
/// Six core stat values (HP, ATK, DEF, SpATK, SpDEF, SPD).
/// Range: 0–31 for IVs, 0–255 for EVs.
/// </summary>
public class StatBlock
{
    [BsonElement("hp")]    public int Hp    { get; set; }
    [BsonElement("atk")]   public int Atk   { get; set; }
    [BsonElement("def")]   public int Def   { get; set; }
    [BsonElement("spatk")] public int SpAtk { get; set; }
    [BsonElement("spdef")] public int SpDef { get; set; }
    [BsonElement("spd")]   public int Spd   { get; set; }
}

/// <summary>
/// Embedded move for a PokemonInstance instance.
/// </summary>
public class PokemonMove
{
    [BsonElement("move_id")]
    public int MoveId { get; set; }

    [BsonElement("move_name")]
    public string MoveName { get; set; } = "Unknown";

    [BsonElement("current_pp")]
    public int CurrentPp { get; set; }
}
